src/adaptation/threephase.jl

```diff
@@ -0,0 +1,83 @@
+######################
+### Mutable states ###
+######################
+
+mutable struct ThreePhaseState
+    i           :: Int
+    window_size :: Int
+    next_window :: Int
+end
+
+################
+### Adapterers ###
+################
+
+# TODO: currently only ThreePhaseAdapter has the filed `n_adapts`. maybe we could unify all
+# Acknowledgement: this adaption settings is mimicing Stan's 3-phase adaptation.
+struct ThreePhaseAdapter <: AbstractCompositeAdapter
+    n_adapts    :: Int
+    pc          :: AbstractPreConditioner
+    ssa         :: StepSizeAdapter
+    init_buffer :: Int
+    term_buffer :: Int
+    state       :: ThreePhaseState
+end
+
+function ThreePhaseAdapter(n_adapts::Int, pc::AbstractPreConditioner, ssa::StepSizeAdapter,
+                           init_buffer::Int=75, term_buffer::Int=50, window_size::Int=25)
+    next_window = init_buffer + window_size - 1
+    return ThreePhaseAdapter(n_adapts, pc, ssa, init_buffer, term_buffer, ThreePhaseState(0, window_size, next_window))
+end
+
+# Ref: https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/windowed_adaptation.hpp
+function is_in_window(tp::ThreePhaseAdapter)
+    return (tp.state.i >= tp.init_buffer) &&
+           (tp.state.i < tp.n_adapts - tp.term_buffer) &&
+           (tp.state.i != tp.n_adapts)
+end
+
+function is_window_end(tp::ThreePhaseAdapter)
+    return (tp.state.i == tp.state.next_window) &&
+           (tp.state.i != tp.n_adapts)
+end
+
+is_final(tp::ThreePhaseAdapter) = tp.state.i == tp.n_adapts
+
+function compute_next_window!(tp::ThreePhaseAdapter)
+    if ~(tp.state.next_window == tp.n_adapts - tp.term_buffer - 1)
+        tp.state.window_size *= 2
+        tp.state.next_window = tp.state.i + tp.state.window_size
+        if ~(tp.state.next_window == tp.n_adapts - tp.term_buffer - 1)
+            next_window_boundary = tp.state.next_window + 2 * tp.state.window_size
+            if (next_window_boundary >= tp.n_adapts - tp.term_buffer)
+                tp.state.next_window = tp.n_adapts - tp.term_buffer - 1
+            end
+        end
+    end
+end
+
+function adapt!(tp::ThreePhaseAdapter, θ::AbstractVector{<:Real}, α::AbstractFloat)
```

Comment:
I refactored the way to decompose end-of-window operations and end-of-adaptations a bit. Would be nice to you double check if this doesn't change any logic for me.

Reply:
sure